File: leetcode/java/leetcode/src/linkedList/_206_reverse_linked_list.java
import java.util.List;

import javax.management.ListenerNotFoundException;

/**
 * https://leetcode-cn.com/problems/reverse-linked-list/
 */
public class _206_reverse_linked_list {
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode newHead = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = newHead;
            newHead = head;
            head = temp;
        }
        return newHead;
    }

    public ListNode reverseList2(ListNode head) {
        if (head == null || head.next == null)  return head;

        ListNode newHead = reverseList2(head.next);
        head.next.next = head;
        head.next = null;

        return newHead;
    }
}

File: leetcode/js/linkedList/_206_reverse_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
 function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
 }

/**
 * https://leetcode-cn.com/problems/reverse-linked-list/
 * @param {ListNode} head
 * @return {ListNode}
 */
 var reverseList = function(head) {
    if (!head || !head.next) return head;
    let newHead = null;
    
    while(head) {
        let temp = head.next;
        head.next = newHead;
        newHead = head;
        head = temp;
    }

    return newHead;
};


var reverseList1 = function(head) {
    if (!head || !head.next) return head;

    let temp = reverseList1(head.next);
    head.next.next = head;
    head.next = null;
    
    return temp;
};

